src/apdu_sign.c

```diff
@@ -578,7 +597,7 @@ void output_end(void) {
             // Doesn't cover the case where dst is src, but different change address, which sets maybe-txn, even
             if((G.maybe_transaction.v.flags & HAS_DESTINATION_ADDRESS) && memcmp(G.u.tx.outputs[0].destination.hash, dest_to_show, 20)) {
                 // If here either, the destination is the signer, but the change address is different, or we need to reject it because of multiple output cells
-                if(dest_is_src || is_second_change) {
+                if((dest_is_src || is_second_change) && N_data.contract_data_type == DISALLOW_CONTRACT_DATA) {
```

Comment:
This is kinda of a hack, but all this self transfer stuff is crazy and I'm (trying to) replace it with something simpler, so this is fine for now in my book.